File: chrome/browser/video_tutorials/internal/android/tutorial_conversion_bridge.cc
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/video_tutorials/internal/android/tutorial_conversion_bridge.h"

#include <memory>
#include <string>

#include "base/android/jni_string.h"
#include "chrome/browser/video_tutorials/internal/jni_headers/TutorialConversionBridge_jni.h"

namespace video_tutorials {

using base::android::ConvertUTF8ToJavaString;

ScopedJavaLocalRef<jobject> CreateJavaTutorialAndMaybeAddToList(
    JNIEnv* env,
    ScopedJavaLocalRef<jobject> jlist,
    const Tutorial& tutorial) {
  return Java_TutorialConversionBridge_createTutorialAndMaybeAddToList(
      env, jlist, static_cast<int>(tutorial.feature),
      ConvertUTF8ToJavaString(env, tutorial.title),
      ConvertUTF8ToJavaString(env, tutorial.video_url.spec()),
      ConvertUTF8ToJavaString(env, tutorial.poster_url.spec()),
      ConvertUTF8ToJavaString(env, tutorial.animated_gif_url.spec()),
      ConvertUTF8ToJavaString(env, tutorial.thumbnail_url.spec()),
      ConvertUTF8ToJavaString(env, tutorial.caption_url.spec()),
      ConvertUTF8ToJavaString(env, tutorial.share_url.spec()),
      tutorial.video_length);
}

ScopedJavaLocalRef<jobject> TutorialConversionBridge::CreateJavaTutorials(
    JNIEnv* env,
    const std::vector<Tutorial>& tutorials) {
  ScopedJavaLocalRef<jobject> jlist =
      Java_TutorialConversionBridge_createTutorialList(env);

  for (const auto& tutorial : tutorials)
    CreateJavaTutorialAndMaybeAddToList(env, jlist, tutorial);

  return jlist;
}

ScopedJavaLocalRef<jobject> TutorialConversionBridge::CreateJavaTutorial(
    JNIEnv* env,
    absl::optional<Tutorial> tutorial) {
  ScopedJavaLocalRef<jobject> jobj;
  if (tutorial.has_value()) {
    jobj = CreateJavaTutorialAndMaybeAddToList(
        env, ScopedJavaLocalRef<jobject>(), tutorial.value());
  }

  return jobj;
}

}  // namespace video_tutorials

File: chrome/browser/video_tutorials/internal/android/java/src/org/chromium/chrome/browser/video_tutorials/bridges/TutorialConversionBridge.java
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.video_tutorials.bridges;

import androidx.annotation.Nullable;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.chrome.browser.video_tutorials.Tutorial;

import java.util.ArrayList;
import java.util.List;

/**
 * Helper class to provide conversion methods between C++ and Java for video tutorials.
 */
@JNINamespace("video_tutorials")
public class TutorialConversionBridge {
    @CalledByNative
    private static List<Tutorial> createTutorialList() {
        return new ArrayList<>();
    }

    @CalledByNative
    private static Tutorial createTutorialAndMaybeAddToList(@Nullable List<Tutorial> list,
            int featureType, String title, String videoUrl, String posterUrl, String animatedGifUrl,
            String thumbnailUrl, String captionUrl, String shareUrl, int videoLength) {
        Tutorial tutorial = new Tutorial(featureType, title, videoUrl, posterUrl, animatedGifUrl,
                thumbnailUrl, captionUrl, shareUrl, videoLength);
        if (list != null) list.add(tutorial);
        return tutorial;
    }
}
